### PR TITLE
Integrate NAP WAF v5

### PR DESCRIPTION
**Pull request by shaun-nx**
### Proposed changes

This PR makes the following changes:
1. Adds 4 new build stages, and make commands, to build the NGINX Ingress Controller with [AppProtect WAF v5](https://docs.nginx.com/nginx-app-protect-waf/v5/releases/)
2. Adds a new Helm values, `controller.appprotect.v5`. Setting this option to `true` will two additional containers that are required for WAF v5 images. See the [AppProtect v5 Kubernetes Admin Guide](https://docs.nginx.com/nginx-app-protect-waf/v5/admin-guide/deploy-on-kubernetes/#waf-services-configuration) for more details.

Resolves https://github.com/nginxinc/kubernetes-ingress/issues/5646

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [ ] I have updated necessary documentation
- [x] I have rebased my branch onto main
- [x] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### CODE REVIEW

File: internal/configs/configurator.go [hunk 1]
```diff
@@ -31,7 +31,7 @@ import (
 
 const (
 	pemFileNameForWildcardTLSSecret = "/etc/nginx/secrets/wildcard" // #nosec G101
-	appProtectBundleFolder          = "/etc/nginx/waf/bundles/"
+	appProtectBundleFolder          = "/etc/app_protect/bundles/"
```
Comment (vepatel) [inline internal/configs/configurator.go:34]: there are some tests in template_test.go and virtualserver_test.go which might need updating as well.

File: charts/nginx-ingress/templates/controller-deployment.yaml [hunk 2]
```diff
@@ -141,6 +141,36 @@ spec:
 {{- if .Values.controller.extraContainers }}
       {{ toYaml .Values.controller.extraContainers | nindent 6 }}
 {{- end }}
+
```
Comment (vepatel) [inline charts/nginx-ingress/templates/controller-deployment.yaml:144]: same is required for daemonset template as well

File: build/Dockerfile [hunk 3]
```diff
@@ -232,6 +260,49 @@ RUN --mount=type=secret,id=nginx-repo.crt,dst=/etc/ssl/nginx/nginx-repo.crt,mode
 # RUN update-ca-certificates
 
 
+############################################# Base image for Debian with NGINX Plus and App Protect WAFv5/DoS #############################################
```
Comment (vepatel) [inline build/Dockerfile:263]: may be we should restrict v5 only to debian:12

File: charts/nginx-ingress/templates/controller-deployment.yaml (hunk 2 above)
Comment (shaun-nx) [reply charts/nginx-ingress/templates/controller-deployment.yaml:144]: I'll re-run tests using Daemonset as well. Nice catch.

File: internal/configs/configurator.go (hunk 1 above)
Comment (shaun-nx) [reply internal/configs/configurator.go:34]: I checked the tests dir and we don’t have any references to the bundles directory in there yet. We will need a test for this for v5 though.

File: build/Dockerfile [hunk 4]
```diff
@@ -70,17 +71,20 @@ ADD --link --chown=101:0 https://cs.nginx.com/static/files/nginx-plus-8.repo ngi
...
 ADD --link --chown=101:0 https://raw.githubusercontent.com/nginxinc/k8s-common/main/files/plus-debian-12.repo debian-plus-12.sources
 ADD --link --chown=101:0 https://raw.githubusercontent.com/nginxinc/k8s-common/main/files/nap-waf-debian-11.repo nap-waf-11.sources
 ADD --link --chown=101:0 https://raw.githubusercontent.com/nginxinc/k8s-common/main/files/nap-dos-debian-11.repo nap-dos-11.sources
+ADD --link --chown=101:0 https://raw.githubusercontent.com/nginxinc/k8s-common/main/files/nap-waf-v5-debian-12.repo nap-waf-v5-12.sources
 ADD --link --chown=101:0 https://raw.githubusercontent.com/nginxinc/k8s-common/main/files/agent-debian-11.repo debian-agent-11.sources
 
 RUN --mount=from=busybox:musl,src=/bin/,dst=/bin/ printf "%s\n" "Acquire::https::pkgs.nginx.com::User-Agent k8s-ic-$IC_VERSION${BUILD_OS##debian-plus}-apt;" >> 90pkgs-nginx \
 	&& printf "%s\n" "user_agent=k8s-ic-$IC_VERSION${BUILD_OS##ubi*plus}-dnf" | tee -a nginx-plus-*.repo \
 	&& sed -i -e "s;%VERSION%;${NGINX_PLUS_VERSION};g" *.sources \
-	&& sed -i -e "y/0/1/" -e "1,8s;/centos;/${NGINX_PLUS_VERSION}/centos;" *.repo \
```
Comment (shaun-nx) [inline build/Dockerfile:83]: Re- add sed and exclude v5 repos

File: build/Dockerfile [hunk 5]
```diff
@@ -350,21 +446,80 @@ RUN --mount=type=secret,id=nginx-repo.crt,dst=/etc/ssl/nginx/nginx-repo.crt,mode
...
+	rm -f /etc/yum.repos.d/app-protect-8.repo; \
+	fi \
+	&& if [ -z "${NAP_MODULES##*dos*}" ]; then \
+	rm -f /etc/yum.repos.d/app-protect-dos-8.repo; \
+	fi \
+	&& dnf clean all
+
+# Uncomment the lines below if you want to install a custom CA certificate
+# COPY build/*.crt /etc/pki/ca-trust/source/anchors/
+# RUN update-ca-trust extract
```
Comment (shaun-nx) [inline build/Dockerfile:501]: Remove this comment

File: charts/nginx-ingress/templates/_helpers.tpl [hunk 6]
```diff
@@ -378,6 +396,37 @@ volumeMounts:
...
+        - all
+  volumeMounts:
+    - name: app-protect-bd-config
+      mountPath: /opt/app_protect/bd_config
+    - name: app-protect-config
+      mountPath: /opt/app_protect/config
+    - name: app-protect-bundles
+      mountPath: /etc/app_protect/bundles
+{{- end}}
+{{- end -}}
```
Comment (oseoin) [inline charts/nginx-ingress/templates/_helpers.tpl:428]: The following need to be configurable:
- image
- imagePullPolicy
- securityContext (for both containers)

File: charts/nginx-ingress/templates/_helpers.tpl [hunk 7]
```diff
@@ -312,6 +312,14 @@ List of volumes for controller.
...
   emptyDir: {}
 {{- end }}
+{{- if .Values.controller.appprotect.v5 }}
+- name: app-protect-bd-config
+  emptyDir: {}
+- name: app-protect-config
+  emptyDir: {}
+- name: app-protect-bundles
+  emptyDir: {}
+{{- end }}
```
Comment (oseoin) [inline charts/nginx-ingress/templates/_helpers.tpl:322]: Needs to support BYO-PVC (we probably don't want to own the PVC, just reference an existing one)

File: charts/nginx-ingress/templates/_helpers.tpl [hunk 8]
```diff
@@ -378,6 +396,37 @@ volumeMounts:
...
 {{- end -}}
 
+{{- define "nginx-ingress.appprotect.v5" -}}
+{{- if .Values.controller.appprotect.v5}}
+- name: waf-enforcer
+  image: private-registry.nginx.com/nap/waf-enforcer:5.2.0
+  imagePullPolicy: Always
+  env:
+    - name: ENFORCER_PORT
+      value: "50000"
```
Comment (shaun-nx) [inline charts/nginx-ingress/templates/_helpers.tpl:406]: Make configurable

File: internal/configs/version1/nginx-plus.tmpl [hunk 9]
```diff
@@ -85,6 +85,10 @@ http {
     {{- end}}
     {{- end}}
 
+    {{- if .AppProtectV5LoadModule}}
+    app_protect_enforcer_address 127.0.0.1:50000;
```
Comment (shaun-nx) [inline internal/configs/version1/nginx-plus.tmpl:89]: make address configurable

File: build/Dockerfile [hunk 10]
```diff
@@ -350,21 +446,80 @@ RUN --mount=type=secret,id=nginx-repo.crt,dst=/etc/ssl/nginx/nginx-repo.crt,mode
 # RUN update-ca-trust extract
 
 
+############################################# Base image for UBI with NGINX Plus and App Protect WAF & DoS #############################################
```
[the remark this answers is not in the document]
Comment (jjngx) [reply build/Dockerfile:449]: +1 for skipping DoS for now

File: charts/nginx-ingress/templates/_helpers.tpl [hunk 11]
```diff
@@ -361,6 +381,16 @@ volumeMounts:
 - mountPath: /var/log/nginx
   name: nginx-log
 {{- end }}
+{{- if .Values.controller.appprotect.v5 }}
+- name: app-protect-bd-config
+  mountPath: /opt/app_protect/bd_config
+- name: app-protect-config
+  mountPath: /opt/app_protect/config
+  # Normally, only the waf-config-mgr should access this dir.
+  # This is here for testing.
```
Comment (oseoin) [inline charts/nginx-ingress/templates/_helpers.tpl:390]: @jjngx Can you confirm if this needs to be removed?

Comment (jjngx) [reply charts/nginx-ingress/templates/_helpers.tpl:390]: yes, please remove

**Comment by oseoin** [comment]
Confirmed Security Monitoring with v5
image

**Review by jjngx — APPROVED** [review]

File: charts/nginx-ingress/values.schema.json [hunk 12]
```diff
@@ -139,6 +148,201 @@
...
+            },
+            "enforcer": {
+              "type": "object",
+              "properties": {
+                "host": {
+                  "type": "string",
+                  "default": "127.0.0.1",
+                  "title": "Port which the App Protect WAF v5 Enforcer process runs on",
+                  "examples": [
+                    50000
```
Comment (pdabelf5) [inline charts/nginx-ingress/values.schema.json:182]:
```suggestion
                    127.0.0.1
```

File: charts/nginx-ingress/values.schema.json [hunk 13]
```diff
@@ -139,6 +148,201 @@
...
+                        "sha256:2710c264e8eaeb663cee63db37b75a1ac1709f63a130fb091c843a6c3a4dc572"
+                      ]
+                    },
+                    "pullPolicy": {
+                      "type": "string",
+                      "default": "IfNotPresent",
+                      "title": "The pullPolicy for the  App Protect WAF v5 Enforcer image",
+                      "allOf": [
+                        {
+                          "$ref": "https://raw.githubusercontent.com/nginxinc/kubernetes-json-schema/master/v1.29.0/_definitions.json#/definitions/io.k8s.api.core.v1.Container/properties/imagePullPolicy"
```
Comment (pdabelf5) [inline charts/nginx-ingress/values.schema.json:231]: did we not update these to 1.30.x

File: internal/configs/configurator.go (hunk 1 above)
Comment (pdabelf5) [inline internal/configs/configurator.go:34]: does this path still work with v4?

File: charts/nginx-ingress/values.schema.json (hunk 13 above)
Comment (oseoin) [reply charts/nginx-ingress/values.schema.json:231]: @pdabelf5 looks like that wasn't merged back to main, I think we should do that separately from this PR?

File: cmd/nginx-ingress/main.go [hunk 14]
```diff
@@ -47,11 +48,13 @@ var (
...
-	appProtectVersionLabel = "appprotect.f5.com/version"
-	agentVersionLabel      = "app.nginx.org/agent-version"
-	appProtectVersionPath  = "/opt/app_protect/RELEASE"
+	nginxVersionLabel        = "app.nginx.org/version"
+	versionLabel             = "app.kubernetes.io/version"
+	appProtectVersionLabel   = "appprotect.f5.com/version"
+	agentVersionLabel        = "app.nginx.org/agent-version"
+	appProtectVersionPath    = "/opt/app_protect/RELEASE"
+	appProtectBundleFolder   = "/etc/nginx/waf/bundles/"
+	appProtectv5BundleFolder = "/etc/app_protect/bundles/"
```
Comment (vepatel) [inline cmd/nginx-ingress/main.go:57]: @oseoin @jjngx does the path for logconf bundle changes based on versions or its same as

File: charts/nginx-ingress/values.yaml [hunk 15]
```diff
@@ -21,9 +21,65 @@ controller:
...
+
+        ## The tag of the App Protect WAF v5 Configuration Manager image.
+        tag: "5.2.0"
+        ## The digest of the App Protect WAF v5 Configuration Manager image.
+        ## If digest is specified it has precedence over tag and will be used instead
+        # digest: "sha256:CHANGEME"
+
+        ## The pull policy for the App Protect WAF v5 Configuration Manager image.
+        pullPolicy: IfNotPresent
+      securityContext:
```
Comment (vepatel) [inline charts/nginx-ingress/values.yaml:75]: @oseoin is this required or can be left empty:
```
  securityContext: {} # Remove curly brackets before adding values
    # allowPrivilegeEscalation: true
    # readOnlyRootFilesystem: true
    # runAsUser: 101 #nginx
    # runAsNonRoot: true
    # capabilities:
    #   drop:
    #   - ALL
    #   add:
    #   - NET_BIND_SERVICE
```

File: cmd/nginx-ingress/main.go (hunk 14 above)
Comment (oseoin) [reply cmd/nginx-ingress/main.go:57]: @vepatel `appProtectv5BundleFolder` is used for v5, otherwise `appProtectBundleFolder` is used

Comment (oseoin) [reply cmd/nginx-ingress/main.go:57]: Just updated naming to make the switch obvious

File: charts/nginx-ingress/values.yaml (hunk 15 above)
Comment (oseoin) [reply charts/nginx-ingress/values.yaml:75]: I think these came from [here](https://docs.nginx.com/nginx-app-protect-waf/v5/admin-guide/deploy-on-kubernetes/) - @jjngx can you confirm if these came from testing?

Comment (jjngx) [reply charts/nginx-ingress/values.yaml:75]: confirmed

**Review by pdabelf5 — APPROVED** [review]

**Review by vepatel — APPROVED** [review]

**Pull request merged**